File: src/main/java/com/cjcrafter/foliascheduler/FoliaCompatibility.java
package com.cjcrafter.foliascheduler;

import org.bukkit.Location;
import org.bukkit.Server;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;

/**
 * The main class of FoliaScheduler, provides methods for getting the server-specific schedulers.
 */
public class FoliaCompatibility {

    private final @NotNull ServerImplementation serverImplementation;

    public FoliaCompatibility(@NotNull Plugin plugin) {
        assertRelocated(plugin);
        ServerImplementation scheduler;

        // Look for the method org.bukkit.Server#isOwnedByCurrentRegion(Location)
        // If such a method exists, use the Folia scheduler. Otherwise, use the Bukkit scheduler.
        try {
            try {
                Server.class.getMethod("isOwnedByCurrentRegion", Location.class);
                scheduler = Class.forName(this.getClass().getPackage().getName() + ".folia.FoliaServer")
                        .asSubclass(ServerImplementation.class)
                        .getConstructor(Plugin.class)
                        .newInstance(plugin);
            } catch (NoSuchMethodException ex) {
                scheduler = Class.forName(this.getClass().getPackage().getName() + ".bukkit.BukkitServer")
                        .asSubclass(ServerImplementation.class)
                        .getConstructor(Plugin.class)
                        .newInstance(plugin);
            }
        } catch (Throwable ex) {
            throw new RuntimeException("Failed to initialize scheduler", ex);
        }

        this.serverImplementation = scheduler;
    }

    /**
     * Gets the server-specific scheduler implementation.
     *
     * @return The server-specific scheduler implementation.
     */
    public @NotNull ServerImplementation getServerImplementation() {
        return serverImplementation;
    }

    private static void assertRelocated(@NotNull Plugin plugin) {
        String originalPackage = String.join(".", "com", "cjcrafter", "scheduler");
        String relocatedPackage = FoliaCompatibility.class.getPackage().getName();
        if (originalPackage.equals(relocatedPackage)) {
            String authors = String.join(", ", plugin.getDescription().getAuthors());
            String possiblePackage = plugin.getClass().getPackage().getName() + ".libs.scheduler";
            plugin.getLogger().warning("The FoliaScheduler lib has not been relocated!");
            plugin.getLogger().warning("The package is still in '" + originalPackage + "'");
            plugin.getLogger().warning("Please relocate the lib to '" + possiblePackage + "'");
            plugin.getLogger().warning("Please warn authors " + authors + " about this issue.");
        }
    }
}

File: src/main/kotlin/com/cjcrafter/scheduler/FoliaCompatibility.kt
package com.cjcrafter.scheduler

import org.bukkit.Location
import org.bukkit.Server
import org.bukkit.plugin.Plugin

/**
 * The main class of FoliaScheduler, provides methods for getting the server-specific schedulers.
 */
class FoliaCompatibility(val plugin: Plugin) {

    /**
     * The server-specific scheduler implementation
     */
    val serverImplementation: ServerImplementation

    init {
        assertRelocated()

        // Look for the method org.bukkit.Server#isOwnedByCurrentRegion(Location)
        // If such a method exists, use the Folia scheduler. Otherwise, use the Bukkit scheduler.
        var scheduler: ServerImplementation
        try {
            try {
                Server::class.java.getMethod("isOwnedByCurrentRegion", Location::class.java)
                scheduler = Class.forName(javaClass.`package`.name + ".folia.FoliaServer")
                    .getDeclaredConstructor(Plugin::class.java)
                    .newInstance(plugin) as ServerImplementation
            } catch (e: NoSuchMethodException) {
                scheduler = Class.forName(javaClass.`package`.name + ".bukkit.BukkitServer")
                    .getDeclaredConstructor(Plugin::class.java)
                    .newInstance(plugin) as ServerImplementation
            }
        } catch (ex: Throwable) {
            throw InternalError("Failed to initialize scheduler", ex)
        }

        this.serverImplementation = scheduler
    }

    private fun assertRelocated() {
        // Cannot use "com.cjcrafter.scheduler" because some relocaters will change it
        val originalPackage = java.lang.String.join(".", "com", "cjcrafter", "scheduler")
        val relocatedPackage = this::class.java.`package`.name
        if (originalPackage == relocatedPackage) {
            val authors = java.lang.String.join(", ", plugin.description.authors)
            val possiblePackage = plugin.javaClass.`package`.name + ".libs.scheduler"
            plugin.logger.warning("The FoliaScheduler lib has not been relocated!")
            plugin.logger.warning("The package is still in '$originalPackage'")
            plugin.logger.warning("Please relocate the lib to '$possiblePackage'")
            plugin.logger.warning("Please warn authors $authors about this issue.")
        }
    }
}